feat (cli): Add command search using Ctrl+r

Aisha630 opened the pull request:
## TLDR

This pull request introduces a new feature to the CLI: command search functionality using `Ctrl+r`. 

## Dive Deeper

The implementation includes:
- Integration of `useReverseSearchCompletion` for command search, enabling incremental search.
- Updates to `InputPrompt` to handle `Ctrl+r` for activating command search and navigating suggestions.
- Enhancements to `SuggestionsDisplay` to support expanded suggestions for long commands.

## Reviewer Test Plan

To validate this change:
1. Pull the branch and run the CLI.
2. Test the command search functionality by pressing `Ctrl+r` and typing a query.
3. Navigate through suggestions using arrow keys and verify expanded suggestions for long commands.


## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | -  | - |
| npx      | - | -  | -  |
| Docker   | - | -  | -  |
| Podman   | -   | -   | -   |
| Seatbelt | -   | -   | -   |

## Linked issues / bugs

Resolves #5391 


https://github.com/user-attachments/assets/9e629647-d389-42a7-a0e2-9380f462f5ff

jacob314 commented:
Let me know when this is ready to review. It is currently marked as (draft) but looks quite close to being ready. Would like to discuss how we handle multiline prompts in the history. There are probably additional things we should consider to make that experience better. @miguelsolorio may have thoughts.
Can't wait for this PR! I've now seen two other PRs be proposed attempting this or a similar feature so I'm pretty sure users will be delighted when this lands.

Aisha630 commented:
/gemini review

Aisha630 commented:
/gemini review

Aisha630 commented:
/gemini review

Aisha630 commented:
/gemini review

Aisha630 commented:
/gemini review

Aisha630 commented:
@jacob314 @miguelsolorio would love to know your ideas about handling multiline commands in history

jacob314 commented:
Can you fix the errors and then I will take a look. Added myself back to the review as I believe @mattKorwel is out this week.

I see "Error: src/ui/components/PrepareLabel.tsx(90,29): error TS1361: 'React' cannot be used as a value because it was imported using 'import type'.
node:internal/errors:984
  const err = new Error(message);"

jacob314 reviewed (commented):
I'm seeing the following strange rendering artifacts.

https://github.com/user-attachments/assets/77d4fa13-02dd-41bd-91fb-15decf7b93ae

Aisha630 commented:
> Another view of the strange things happening when the completion is almost the full string. <img alt="Screenshot 2025-08-28 at 6 09 24 PM" width="1108" height="124" src="https://private-user-images.githubusercontent.com/1226812/483407168-d060afe8-9fec-4aa8-8780-a3e852b65f0c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY1OTY1MTAsIm5iZiI6MTc1NjU5NjIxMCwicGF0aCI6Ii8xMjI2ODEyLzQ4MzQwNzE2OC1kMDYwYWZlOC05ZmVjLTRhYTgtODc4MC1hM2U4NTJiNjVmMGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MzBUMjMyMzMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmVkNTY2YjY0ZTBlZDVhZmM2ZWZlODY3NDlkMWZmYzU3OGUxZmY4MDcyMjMzZDVkMGVlODFmYjNhOWQ1NzZhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4Vy4nYOeAASluWxh6JKEd2T8mdTohTeOECXjeQ6rdtQ">
> 
> Same text when selected to show how the blank text is just a black background so it is likely text that should have the background color but that is not working. <img alt="Screenshot 2025-08-28 at 6 09 24 PM" width="1108" height="124" src="https://private-user-images.githubusercontent.com/1226812/483407405-e25415b5-fb2c-4d68-aa17-1b683a8e3c29.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY1OTY1MTAsIm5iZiI6MTc1NjU5NjIxMCwicGF0aCI6Ii8xMjI2ODEyLzQ4MzQwNzQwNS1lMjU0MTViNS1mYjJjLTRkNjgtYWExNy0xYjY4M2E4ZTNjMjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MzBUMjMyMzMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODkwMzViZGNiZDY4Y2E0MTIwNGQ3NGE3MzMyYzU3ZGQzYzU1YzVlMjViM2VhYTkzMWNhMGE0MTRkMmVjMGIxYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.jbvk17Up363J_ykNANnhmSx2FzNGSescGrJ92sWtne4">
> 
> unclear why the text width is like it is. seems like we are not quite using the box model right in ink for it or something. the completions are set to have a width of 20 chars max yet that isn't respected (and shouldn't be because that would look silly).
> 
> I'll approve once this is addressed. Sorry this review has dragged on so long.

I have fixed this as well. The width was weird since this was a slashCommand and that had a particular width set. However this should not be this way if reverse search mode is active so I have changed this as well now

jacob314 reviewed (commented):
All the changes look good now. The one thing that is missing is test coverage. Please add some basic snapshot based tests for .tsx files changed so we don't accidentally regress this functionality. Cases such as the expand and collapse suggestion functionality are also worth testing.
The only UX concern I have is  line breaking sometimes changes when lines are focused on. This is due to the arrows shown to expand and collapse entries. @miguelsolorio would be interested on your thoughts on this. Potentially we should insert a blank spot suitable for the arrows even when a line isn't focused to avoid relayout or just omit the arrows.
https://github.com/user-attachments/assets/9956b959-9b44-4b3a-ac43-8c2b0c665ffd
We can resolve that as a follow up so don't want to block landing this PR for that UI polish.

Aisha630 commented:
> All the changes look good now. The one thing that is missing is test coverage. Please add some basic snapshot based tests for .tsx files changed so we don't accidentally regress this functionality. Cases such as the expand and collapse suggestion functionality are also worth testing. The only UX concern I have is line breaking sometimes changes when lines are focused on. This is due to the arrows shown to expand and collapse entries. @miguelsolorio would be interested on your thoughts on this. Potentially we should insert a blank spot suitable for the arrows even when a line isn't focused to avoid relayout or just omit the arrows. https://github.com/user-attachments/assets/9956b959-9b44-4b3a-ac43-8c2b0c665ffd We can resolve that as a follow up so don't want to block landing this PR for that UI polish.

Done. I have added tests for this functionality
